### PR TITLE
Switch to ExAws

### PR DESCRIPTION
**Pull request by mschae**
We should use @benwilson512's excellent ExAws library for handling S3 uploads instead of mini_s3.

The primary reason being that mini_s3 currently won't work with OTP18 (or rebar3 whatever it is).

### CODE REVIEW

**Comment by mschae** [comment]
This is **WIP** until the ExAws changes around S3 are fixed.

File: config/config.exs [hunk 1]
```diff
@@ -28,6 +26,15 @@ config :hex_web, HexWeb.Repo,
   adapter: Ecto.Adapters.Postgres,
   extensions: [{HexWeb.JSON.Extension, library: Poison}]
 
+config :ex_aws,
+  access_key_id:    [{:system, "HEX_S3_ACCESS_KEY"}],
+  secret_acces_key: [{:system, "HEX_S3_SECRET_KEY"}]
```
Comment (benwilson512) [inline config/config.exs]: This should be `secret_access_key`

**Comment by mschae** [comment]
@benwilson512: Please give this a fist pass.

@ericmj: This is ready for a first review while @benwilson512 wraps up the S3 integration and releases a new version of ex_aws.

**Comment by mschae** [comment]
Didn't see @benwilson512's PR. Sorry about that. Merged now, looks even better. :)

File: mix.exs [hunk 2]
```diff
@@ -28,7 +28,9 @@ defmodule HexWeb.Mixfile do
      {:earmark,   "~> 0.1"},
      {:gen_smtp,  "~> 0.9.0"},
      {:bcrypt,    github: "opscode/erlang-bcrypt"},
-     {:mini_s3,   github: "ericmj/mini_s3", branch: "hex-fixes"}
+     {:httpoison, "~> 0.7.0"},
+     {:sweet_xml, "~> 0.2.1"},
+     {:ex_aws,    github: "CargoSense/ex_aws", branch: "finish-s3"}
```
Comment (benwilson512) [inline mix.exs]: Will do a release shortly and give the correct hex version.

File: lib/hex_web/store/s3.ex [hunk 3]
```diff
@@ -103,56 +86,16 @@ defmodule HexWeb.Store.S3 do
...
-
-  defp config do
-    access_key = Application.get_env(:hex_web, :s3_access_key) |> String.to_char_list
-    secret_key = Application.get_env(:hex_web, :s3_secret_key) |> String.to_char_list
-    url = Application.get_env(:hex_web, :s3_url) |> String.to_char_list
-    s3_config(access_key_id: access_key, secret_access_key: secret_key, url: url)
+    Application.get_env(:hex_web, bucket)
+    |> S3.stream_objects!(prefix: prefix)
+    |> Stream.map(&Map.get(&1, :key))
   end
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex]: I'm particularly happy with this change :)

Comment (mschae) [inline lib/hex_web/store/s3.ex]: You and me both. :)

File: mix.exs (hunk 2 above)
Comment (benwilson512) [inline mix.exs]: 0.4.0 is out. This should now be
`{:ex_aws, "~> 0.4.0"}`

**Comment by mschae** [comment]
I just used hex to update a hex dependency.

**Comment by mschae** [comment]
This is good for a final review.

File: lib/hex_web/store/s3.ex [hunk 4]
```diff
@@ -1,36 +1,24 @@
...
 
   def get_logs(key) do
-    bucket = Application.get_env(:hex_web, :logs_bucket) |> String.to_char_list
-    key = String.to_char_list(key)
-    result = retry(fn -> :mini_s3.get_object(bucket, key, [], config()) end)
-    result[:content]
+    Application.get_env(:hex_web, :logs_bucket)
+    |> S3.get_object!(key)
+    |> Map.get(:body)
```
Comment (mschae) [inline lib/hex_web/store/s3.ex]: Wondering wether this should be `Dict.fetch!(:body)` in the spirit of raising when missing...

Comment (benwilson512) [inline lib/hex_web/store/s3.ex]: The get_object! Ensures that get_object returned {:ok, resp} indicating a
2XX status code.  resp in such a situation is guaranteed to have a body
key.

**Comment by ericmj** [comment]
This looks great. Can you rebase away the merges, the PR merge should stay.

**Comment by mschae** [comment]
I think I caught another error. Sorry @ericmj but I have no idea how to do that. :(

**Comment by benwilson512** [comment]
What's the nature of the error?

**Comment by mschae** [comment]
It looks like you re-introduced the `:secret_acces_key` :) https://github.com/hexpm/hex_web/commit/62a790200bec642b448028930bda2c408cd69fbd

File: config/config.exs [hunk 5]
```diff
@@ -28,6 +26,15 @@ config :hex_web, HexWeb.Repo,
   adapter: Ecto.Adapters.Postgres,
   extensions: [{HexWeb.JSON.Extension, library: Poison}]
 
+config :ex_aws,
+  access_key_id:     {:system, "HEX_S3_ACCESS_KEY"},
+  secret_access_key: {:system, "HEX_S3_SECRET_KEY"}
```
Comment (benwilson512) [inline config/config.exs]: @mschae this line looks ok to me.

**Comment by ericmj** [comment]
I will rebase and merge manually if that's ok with both of you?

**Comment by benwilson512** [comment]
@ericmj fire away!

File: config/config.exs (hunk 5 above)
Comment (mschae) [inline config/config.exs]: It does now. :)

**Comment by mschae** [comment]
@ericmj thanks, sounds good.

**Comment by ericmj** [comment]
I couldn't fix the history, too many merges. In the future try to avoid creating merge commits, always rebase when merging other branches into your PR branch.

**Comment by ericmj** [comment]
Thank you so much for this @mschae and @benwilson512!

**Comment by jcspencer** [comment]
This is going to make closing #55 much easier! Thank a ton @mschae and @benwilson512!

**Pull request merged**